### DISCUSSION TITLE
Self-referencing NFACE element ?

### DISCUSSION THREAD

**Discussion by awoimbee** [CGNS/CGNS#373] [CGNStalk]
Hi,

I had multiple CGNS files that made [VTK](https://gitlab.kitware.com/vtk/vtk) crash when reading them. So I wrote my own parser on top of the MLL.
This parser works great, but I've had issues with some NFACE elements. Some polyhedrons are referencing other polyhedrons instead of polygons.

---

Example:

Inside a zone, the last non-NFACE element ends at `386_209_602`, I have a single NFACE element that spans `386_209_603` to `510_885_434`. The connectivity array of the NFACE element contains the value `386_217_145`.

```rust
    #[test]
    fn invalid_polyhedron() {
        let f = File::new("./segfault-volume2.cgns", cgns_rust::file::OpenFileMode::Read);
        let b = f.iter_bases().next();
        let z = b.iter_zones().next();
        let ngon: Vec<_> = z.iter_elements().filter(|e| e.elem_type != ElementType_t::NFACE_n).collect();
        let nface: Vec<_> = z.iter_elements().filter(|e| e.elem_type == ElementType_t::NFACE_n).collect();
        assert_eq!(nface.len(), 1);
        let max_ngon = ngon.iter().map(|e| e.elem_end).max();
        let conn = nface[0].read_connectivity(); // basically cg_poly_elements_read
        for c in conn.connectivity {
            assert!(c < max_ngon, "index {} out of range for 2d elems (>={}), nface elem spans: {} - {}", c, max_ngon, nface[0].elem_start, nface[0].elem_end);
        }
    }
``` 

```
index 386217145 out of range for 2d elems (>=386209602), nface elem spans: 386209603 - 510885434)
```

How am I supposed to handle that ? Up until now I have removed these "invalid" cells from the final result, but is there a better solution ?

**Comment by mennodeij** [comment]
> Some polyhedrons are referencing other polyhedrons instead of polygons.

How did that happen? Is the CNGS file valid according to cgnscheck? 
If they are valid and are crashing the VTK reader, please report an issue about this at https://gitlab.kitware.com/vtk/vtk/issues - preferably with a small CGNS file that leads to a crash.

**Reply by awoimbee** [reply]
I have no idea how the exports are done, these files are sent to me and I need to be able to read them, even if they are slightly invalid. They are also quite big (>=30GB, debugging is a nightmare).
Most files are `CGNS 3.21` from `STAR-CCM+ 2020.2`.

A heavily redacted and shortened cgnscheck output looks like:
```
$ cgnscheck segfault-volume.cgns
reading CGNS file segfault-volume.cgns

reading base "Base"
reading zone "01-[redacted]"
  reading element set "[redacted]"
  [...]
  skipping volume face checking since there were 3 errors
reading zone "02-Fake_ Name"
  reading element set "[redacted]"
  building volume faces hash table...
  finding exterior nodes...
[...]

checking base "Base"
checking base iterative data "TimeIterValues"
  checking "TimeValues"
  checking "IterationValues"

checking zone "01-[redacted]"
WARNING:No family name declared for zone "01-[redacted]".It is a recommended practice to associate one.
  checking coordinates "GridCoordinates"
    checking coordinate "CoordinateX"
WARNING:exponents not given
    checking coordinate "CoordinateY"
WARNING:exponents not given
    checking coordinate "CoordinateZ"
WARNING:exponents not given
  checking elements
  checking element set "[redacted]"
  [...]
ERROR:3 polyhedra faces are not valid face elements
  checking boundary conditions
  checking BC "[redacted]"
    checking BC interface
  [...]
  checking zone grid connectivty "Zcon_01-[redacted]"
CGNSlib ERROR:cg_conn_info:cg_conn_info:donor zone 02-Fake: name does not exist
```

**Comment by mennodeij** [comment]
Well, it looks like the STAR-CCM+ CGNS writing is at fault here. It is not trivial to use the polyhedra writing correctly with the correct indices to refer to faces. Seeing that it is version 3.21 these files are coming probably from an older version of STAR-CCM, so this may be fixed in newer versions of STAR-CCM. 
Without more information it will be difficult to reverse engineer the mistake and work around it I'm afraid.

**Reply by awoimbee** [reply]
Well, I guess I'll keep removing invalid cells.

Regarding the CGNS version:
Maybe StarCCM+ had issues with CGNS due to the v3.4 debacle. I have files coming from `STAR-CCM+ 2021.2.1` with `CGNSLibraryVersion 3.4` (in reality v3.4.1 🤢). The latest documentation (from 10/2022) states:
```
CGNS Version (Read-Only)
Specifies the version of the CGNS standard in use. Simcenter STAR-CCM+ currently supports version 3.2.1
```

**Reply by mennodeij** [reply]
Yes they did, we have also seen faulty files from star-ccm that was due to being written with 3.4.1 (with the reverted old layout) but having CGNS version 3400 in the file, and therefore interpreted as new layout. One thing that worked was to manually edit the version in the file using the cgnsview utility (I guess you could also directly edit the hdf5 file, much faster with large files).
